vortex-array/src/array/primitive/compute/filter_indices.rs

```diff
@@ -0,0 +1,246 @@
+use itertools::Itertools;
+use vortex_dtype::{match_each_native_ptype, NativePType};
+use vortex_error::{vortex_bail, VortexResult};
+use vortex_expr::expressions::{Disjunction, Predicate, Value};
+use vortex_expr::operators::Operator;
+
+use crate::array::primitive::PrimitiveArray;
+use crate::compute::filter_indices::FilterIndicesFn;
+use crate::validity::Validity;
+use crate::{Array, ArrayTrait, IntoArray};
+
+impl FilterIndicesFn for PrimitiveArray {
+    fn filter_indices(&self, predicate: &Disjunction) -> VortexResult<Array> {
+        let conjunction_indices = predicate
+            .conjunctions
+            .iter()
+            .map(|conj| {
+                MergeOp::All(
+                    conj.predicates
+                        .iter()
+                        .map(|pred| indices_matching_predicate(self, pred).unwrap())
```

Comment:
This is a good point -- initially I wanted to avoid doing pairwise reductions in favor of a single-pass, but the allocations/vec overhead here might outweigh that benefit anyway. I'll rewrite this to be fully lazy and use bitmaps instead of vecs.

The one thing I'm not sure of here is whether we have a 64-bit bitmap in the rust croaring crate -- at first glance I didn't see one, will take a closer look tomorrow.

Reply:
So it's not actually that hot. It's an allocation per predicate, so an expression of (X > 2 &  X < 10) is two predicates, regardless of how big the array X is.

With the right bitset utility, you should be able to mutate-in-place instead of allocating a third bitset for the result. But without benchmarking, it's hard to intuit which will perform better.